ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt

```diff
@@ -0,0 +1,178 @@
+import com.google.android.ground.model.mutation.LocationOfInterestMutation.Companion.builder
+import com.google.android.ground.model.mutation.Mutation
+import com.google.android.ground.model.mutation.Mutation.SyncStatus
+import com.google.android.ground.persistence.local.LocalDataStore
+import com.google.android.ground.persistence.local.LocalValueStore
+import com.google.android.ground.persistence.local.room.models.MutationEntitySyncStatus
+import com.google.android.ground.persistence.remote.NotFoundException
+import com.google.android.ground.persistence.remote.RemoteDataEvent
+import com.google.android.ground.persistence.remote.RemoteDataStore
+import com.google.android.ground.persistence.sync.DataSyncWorkManager
+import com.google.android.ground.persistence.uuid.OfflineUuidGenerator
+import com.google.android.ground.rx.annotations.Cold
+import com.google.android.ground.system.auth.AuthenticationManager
+import com.google.common.collect.ImmutableList
+import com.google.common.collect.ImmutableSet
+import io.reactivex.*
+import java8.util.Optional
+import timber.log.Timber
+import java.util.*
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Coordinates persistence and retrieval of [LocationOfInterest] instances from remote, local,
+ * and in memory data stores. For more details on this pattern and overall architecture, see
+ * https://developer.android.com/jetpack/docs/guide.
+ */
+@Singleton
+class LocationOfInterestRepository @Inject constructor(
+    private val localDataStore: LocalDataStore,
+    private val localValueStore: LocalValueStore,
+    private val remoteDataStore: RemoteDataStore,
+    private val surveyRepository: SurveyRepository,
+    private val dataSyncWorkManager: DataSyncWorkManager,
+    private val authManager: AuthenticationManager,
+    private val uuidGenerator: OfflineUuidGenerator
+) {
+    /**
+     * Mirrors locations of interest in the specified survey from the remote db into the local db when
+     * the network is available. When invoked, will first attempt to resync all locations of interest
+     * from the remote db, subsequently syncing only remote changes. The returned stream never
+     * completes, and subscriptions will only terminate on disposal.
+     */
+    fun syncLocationsOfInterest(survey: Survey): @Cold Completable {
+        return remoteDataStore
+            .loadLocationsOfInterestOnceAndStreamChanges(survey)
+            .flatMapCompletable { updateLocalLocationOfInterest(it) }
+    }
+
+    // TODO: Remove "location of interest" qualifier from this and other repository method names.
+    private fun updateLocalLocationOfInterest(event: RemoteDataEvent<LocationOfInterest>): @Cold Completable {
+        return when (event.eventType) {
+            RemoteDataEvent.EventType.ENTITY_LOADED, RemoteDataEvent.EventType.ENTITY_MODIFIED ->
+                event.entity.getOrNull()
+                    ?.let { localDataStore.mergeLocationOfInterest(it) }
```

Comment:
Instead of chaining `getOrNull()` + `?.let`, consider using `fold()`:
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/fold.html